### PR TITLE
Use std::filesystem

### PR DESCRIPTION
**Pull request by chfast**
This uses `std::filesystem` if available or a third party implementation of the filesystem API placed in the `thirdparty/` dir.

There is a single simple use of `fs::path` for demonstration.

On GCC8 might require linking with `stdc++-fs` when more complicated API is used.

### CODE REVIEW

File: src/openrct2/core/Path.cpp [hunk 1]
```diff
@@ -22,6 +22,31 @@
...
+#if defined(_MSC_VER)
+#    define HAVE_STD_FILESYSTEM 1
+#elif defined(__APPLE__)
+#    define HAVE_STD_FILESYSTEM 0
+#elif defined(__has_include)
+#    if __has_include(<filesystem>)
+#        define HAVE_STD_FILESYSTEM 1
+#    else
+#        define HAVE_STD_FILESYSTEM 0
+#    endif
```
Comment (tupaschoal) [inline src/openrct2/core/Path.cpp]: Why is this necessary? Which case is it trying to cover other than the two above? Just trying to understand

Comment (tupaschoal) [reply src/openrct2/core/Path.cpp]: This kind of got stale while I was doing it.

File: src/openrct2/core/Path.cpp [hunk 2]
```diff
@@ -22,6 +22,31 @@
 
 #include <iterator>
 
+#if defined(_MSC_VER)
+#    define HAVE_STD_FILESYSTEM 1
```
[the remark this answers is not in the document]
Comment (Gymnasiast) [reply src/openrct2/core/Path.cpp]: It's convention.

[the remark this answers is not in the document]
Comment (chfast) [reply src/openrct2/core/Path.cpp]: I don't care, can change it to any name. I should also `#undef` it after we include the correct header.

File: src/openrct2/core/Path.cpp (hunk 1 above)
Comment (chfast) [reply src/openrct2/core/Path.cpp]: I have added comments. Which part is confusing?

File: src/openrct2/core/Path.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Gymnasiast) [reply src/openrct2/core/Path.cpp]: I'd wait for other team members to reply before you do anything. (Note that tupaschoal is a contributor, but not a team member.)

**Comment by duncanspumpkin** [comment]
Just a thought but can't we just simplify this to file system is used in all cases apart from xcode. We are usually pretty aggressive at supporting the latest standards. It should be blatantly obvious that it is a hack just for xcode.

**Comment by chfast** [comment]
> Just a thought but can't we just simplify this to file system is used in all cases apart from xcode. We are usually pretty aggressive at supporting the latest standards. It should be blatantly obvious that it is a hack just for xcode.

You need to upgrade your CI toolchains first. Some of them are failing. Here is an earlier build that shows failures: https://travis-ci.org/OpenRCT2/OpenRCT2/builds/613695998

All Linux builds are failing due to using GCC 7 (also the one with Clang 5 which will use the stdc++ lib from GCC 7). You would need to upgrade at least to GCC 8, but I'd recommend GCC 9 (no need to link against stdc++-fs).

**Comment by IntelOrca** [comment]
`std::filesystem` will replace things like `Path.cpp` so the plan is to use `std::filesystem` in our actual core source code. This means we probably want some sort of common header that all other source code can include which do the #ifdef macos etc. boilerplate code.

**Comment by chfast** [comment]
> `std::filesystem` will replace things like `Path.cpp` so the plan is to use `std::filesystem` in our actual core source code. This means we probably want some sort of common header that all other source code can include which do the #ifdef macos etc. boilerplate code.

Yes, that make sense. Where would you like to put it?

**Comment by IntelOrca** [comment]
We have done a very similar thing for `std::optional`, so I suggest copying that:
https://github.com/OpenRCT2/OpenRCT2/blob/develop/src/openrct2/core/Optional.hpp

E.g. `FileSystem.hpp`.

**Comment by chfast** [comment]
> E.g. `core/FileSystem.hpp`.

Done.

**Comment by IntelOrca** [comment]
It is fine, we don't actually need the `GetExtension` method anymore as everything can directly use the filesystem API but this PR gets things started.

**Comment by chfast** [comment]
> It is fine, we don't actually need the `GetExtension` method anymore as everything can directly use the filesystem API but this PR gets things started.

I just wanted any use case for demonstration. I also re-implemented `Path::Combine()` with it, but that caused some issues so it might not be so straight forward. Also, there is also API for C-strings in Path what will require more work outside.

**Comment by janisozaur** [comment]
Hey @chfast can you check https://github.com/OpenRCT2/OpenRCT2/issues/10283 ? Looks like a regression from this PR
